readthedocs/integrations/models.py

```diff
@@ -0,0 +1,145 @@
+        :type payload: dict
+        """
+        request_payload = payload
+        if request_payload is None:
+            request_payload = getattr(req, 'data', None)
+            if req.content_type != 'application/json':
+                # Here, request_body can be a dict or a MergeDict. Probably best
+                # to normalize everything first
+                try:
+                    request_payload = dict(request_payload.items())
+                except AttributeError:
+                    pass
+        try:
+            request_body = json.dumps(request_payload, sort_keys=True)
+        except TypeError:
+            request_body = str(request_payload)
+        # This is the rawest form of request header we have, the WSGI
+        # headers. HTTP headers are prefixed with `HTTP_`, which we remove,
+        # and because the keys are all uppercase, we'll normalize them to
+        # title case-y hyphen separated values.
+        request_headers = dict(
+            (key[5:].title().replace('_', '-'), str(val))
+            for (key, val) in req.META.items()
+            if key.startswith('HTTP_')
+        )
+        request_headers['Content-Type'] = req.content_type
+
+        response_payload = resp.data if hasattr(resp, 'data') else resp.content
+        try:
+            response_body = json.dumps(response_payload, sort_keys=True)
+        except TypeError:
+            response_body = str(response_payload)
+        response_headers = dict(resp.items())
+
+        fields = {
+            'status_code': resp.status_code,
+            'request_headers': request_headers,
+            'request_body': request_body,
+            'response_body': response_body,
+            'response_headers': response_headers,
+        }
+        fields['related_object'] = related_object
+        self.delete_limit(related_object, limit=9)
+        return self.create(**fields)
+
+    def delete_limit(self, related_object, limit=10):
+        queryset = self.filter(
+            content_type=ContentType.objects.get(
+                app_label=related_object._meta.app_label,
+                model=related_object._meta.model_name,
+            ),
+            object_id=related_object.pk
+        )
+        for txn in queryset[limit:]:
+            txn.delete()
+
+
+class HttpTransaction(models.Model):
+
+    """HTTP request/response transaction"""
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    related_object = GenericForeignKey('content_type', 'object_id')
+
+    date = models.DateTimeField(_('Date'), auto_now_add=True)
+
+    request_headers = JSONField(_('Request headers'))
+    request_body = models.TextField(_('Request body'))
+
+    response_headers = JSONField(_('Request headers'))
+    response_body = models.TextField(_('Response body'))
```

Comment:
We can move clean up away from the object creation, perhaps into a celery beat task or something, but this at least gets us started.